runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -5,16 +5,560 @@
+
+    private static final String XWALK_CORE_MARKET_URL =
+            "market://details?id=" + XWalkCoreWrapper.XWALK_APK_PACKAGE;
+
+    private static final String XWALK_CORE_DOWNLOAD_URL =
+            "http://10.0.2.2/" + XWALK_CORE_APK; // For test only
+
+    private DownloadTask mDownloadTask;
+    private Dialog mActiveDialog;
+    private XWalkCoreWrapper mCoreWrapper;
+    private XWalkLibraryDelegate mLibraryDelegate;
+
+    private boolean mIsXWalkReady;
+    private boolean mIsShowing;
+    private boolean mIsInitializingLibrary;
+    private LinkedList<Object> mReservedObjects;
+    private LinkedList<ReflectMethod> mReservedMethods;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        mDownloadTask = null;
+        mActiveDialog = null;
+        mCoreWrapper = null;
+        mLibraryDelegate = null;
+
+        mIsXWalkReady = false;
+        mIsShowing = false;
+        mIsInitializingLibrary = false;
+        mReservedObjects = new LinkedList<Object>();
+        mReservedMethods = new LinkedList<ReflectMethod>();
+
+        XWalkCoreWrapper.reset(mCoreWrapper, this);
+        XWalkCoreWrapper.check();
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        mIsShowing = true;
+
+        XWalkCoreWrapper.reset(mCoreWrapper, this);
+        if (!isXWalkReady() && !isShowingDialog() && !mIsInitializingLibrary) {
+            XWalkCoreWrapper.check();
+        }
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        mIsShowing = false;
+    }
 
     @Override
     public Resources getResources() {
         return getApplicationContext().getResources();
     }
+
+    @Override
+    public final void onObjectInitFailed(Object object) {
+        Log.d("XWalkActivity", "Reserve object: " + object.getClass());
+        mReservedObjects.add(object);
+    }
+
+    @Override
+    public final void onMethodCallMissed(ReflectMethod method) {
+        Log.d("XWalkActivity", "Reserve method: " + method.toString());
+        mReservedMethods.add(method);
+    }
+
+    @Override
+    public final void onXWalkLibraryStartupError(LibraryStatus status, Throwable e) {
+        onXWalkStartupError(status, e);
+    }
+
+    @Override
+    public final void onXWalkLibraryRuntimeError(LibraryStatus status, Throwable e) {
+        onXWalkRuntimeError(status, e);
+    }
+
+    @Override
+    public final void onXWalkLibraryCancelled() {
+        finish();
+    }
+
+    @Override
+    public final void onXWalkLibraryMatched() {
+        Log.d("XWalkActivity", "XWalk library matched");
+        XWalkCoreWrapper.init();
+        mCoreWrapper = XWalkCoreWrapper.getInstance();
+        mLibraryDelegate = XWalkLibraryDelegate.newInstance(this, this);
+
+        mIsInitializingLibrary = true;
+        if (!mLibraryDelegate.initLibrary()) {
+            Assert.fail("XWalk library initialization failed");
+        }
```

Comment:
Should release build keep running when initLibary() failed? If not the case, please do not use Assert